Contextualize

jrchudy opened the pull request:
Contextualize.record function written with unit tests. _referenceCopy also written. Interesting way I found to create an Obj without using it's constructor and then copy all of the data to it.

Assigning to @chiragsanghvi, @RFSH, @howdyjessie for review.

chiragsanghvi commented on js/reference.js:
```diff
@@ -253,8 +256,18 @@ var ERMrest = (function(module) {
+                newRef._columns = [];
+                for (var i = 0; i < columnOrders.length; i++) {
+                    var column = columnOrders[i];
+                    if (source._columns.indexOf(column)) {
```
This will not consider the column which is at index **0** in the `source._columns` array. Instead this statement should check for non **-1** index instead of positive index.

```javascript
if (source._columns.indexOf(column) != -1) {
```

chiragsanghvi commented on js/reference.js:
```diff
@@ -496,8 +509,7 @@ var ERMrest = (function(module) {
     function _referenceCopy(source) {
-        // TODO: make a (copy-on-write) copy of the source reference
-        return source; // TODO
+        return Object.assign(Object.create(Reference.prototype), source);
```
@jrchudy 

This method of creating objects without invoking their constructor is really good. 

Should we make **copy** part of some utility function, instead of declaring it in this class ? So we can use it in other classes too.

For example
```javascript
module._copy = function(source, destClass) {
    var obj = {};
   
    // destClass parameter is optional, if you just want to clone an existing object
    if (destClass) obj = Object.create(destClass.prototype);
   
    return Object.assign(obj, source);
}; 
```

RFSH replied:
Since we are supporting IE, we should include the [Polyfill](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Polyfill) for `Object.assign`.

RFSH commented on test/specs/reference/tests/01.reference.js:
```diff
@@ -56,6 +56,14 @@ exports.execute = function (options) {
+
+                expect(recordReference).not.toBe(reference);
+                expect(recordReference.columns.length).not.toBe(reference.columns.length);
+                expect(recordReference.columns.length).toBe(2);
```
The test case should also check for the actual columns:
```javascript
```

RFSH replied:
The test case should also check for the actual columns:
```javascript
var columns = Array.prototype.map.call(recordReference.columns, function(column){
  return column.name;
});
expect(columns).toEqual(["name", "value"])
```
Also we should check that other properties of the reference are not changed.

jrchudy commented:
@chiragsanghvi and @RFSH, I updated the PR based on your comments that were above. I removed the use of `Object.assign()` in favor of using our `_clone` utility function. I updated the test case as well.

howdyjessie commented:
I've got nothing to add here; it looks good to me. :)
